hudi-client/src/main/java/org/apache/hudi/metadata/HoodieMetadataImpl.java

```diff
@@ -0,0 +1,1003 @@
+        .withCompactionConfig(writeConfig.getMetadataCompactionConfig().orElseGet(() -> getDefaultCompactionConfig()));
+
+    if (writeConfig.isMetricsOn()) {
+      HoodieMetricsConfig.Builder metricsConfig = HoodieMetricsConfig.newBuilder()
+          .withReporterType(writeConfig.getMetricsReporterType().toString())
+          .on(true);
+      switch (writeConfig.getMetricsReporterType()) {
+        case GRAPHITE:
+          metricsConfig.onGraphitePort(writeConfig.getGraphiteServerPort())
+                       .toGraphiteHost(writeConfig.getGraphiteServerHost())
+                       .usePrefix(writeConfig.getGraphiteMetricPrefix());
+          break;
+        case JMX:
+          metricsConfig.onJmxPort(writeConfig.getJmxPort())
+                       .toJmxHost(writeConfig.getJmxHost());
+          break;
+        case DATADOG:
+          // TODO:
+          break;
+        case CONSOLE:
+        case INMEMORY:
+          break;
+        default:
+          throw new HoodieMetadataException("Unsupported Metrics Reporter type " + writeConfig.getMetricsReporterType());
+      }
+
+      builder.withMetricsConfig(metricsConfig.build());
+    }
+
+    return builder.build();
+  }
+
+  private HoodieCompactionConfig getDefaultCompactionConfig() {
+    return HoodieCompactionConfig.newBuilder()
+            .withAutoClean(true)
+            .withInlineCompaction(true)
+            .withCleanerPolicy(HoodieCleaningPolicy.KEEP_LATEST_COMMITS)
+            .archiveCommitsWith(24, 30)
+            .withMaxNumDeltaCommitsBeforeCompaction(24)
+            .build();
+  }
+
+  public HoodieWriteConfig getWriteConfig() {
+    return config;
+  }
+
+  /**
+   * Initialize the metadata table if it does not exist. Update the metadata to bring it in sync with the file system.
+   *
+   * This can happen in two ways:
+   * 1. If the metadata table did not exist, then file and partition listing is used
+   * 2. If the metadata table exists, the instants from active timeline are read in order and changes applied
+   *
+   * The above logic has been chosen because it is faster to perform #1 at scale rather than read all the Instants
+   * which are large in size (AVRO or JSON encoded and not compressed) and incur considerable IO for de-serialization
+   * and decoding.
+   */
+  private void initAndSyncMetadataTable() throws IOException {
+    ValidationUtils.checkState(!readOnly, "Metadata table cannot be initialized in readonly mode");
+
+    final Timer.Context context = this.metrics.getMetadataCtx(INITIALIZE_STR);
+
+    HoodieTableMetaClient datasetMetaClient = new HoodieTableMetaClient(hadoopConf, basePath);
+    FileSystem fs = FSUtils.getFs(metadataBasePath, hadoopConf);
+    boolean exists = fs.exists(new Path(metadataBasePath));
+
+    if (!exists) {
+      // Initialize for the first time by listing partitions and files directly from the file system
+      initFromFilesystem(datasetMetaClient);
```

Comment:
I think I also posted this as a general design question. Repeating here. how does this interplay with already in-flight compactions/cleaning. that may not have metadata table enabled?